### PR TITLE
NO-REF: Improving report scalability

### PR DESCRIPTION
**Pull request by kylevillegas93**
### Performance improvements
- Minimizing calls to DRB database by pulling all publisher project-related data from the DRB db beforehand, normalizing said data to Counter 5 formatting, and hashing for easier search. This is all done before we start processing access logs and generating reports. 
- Aggregating relevant logs before generating reports

### Changes requested by business analysts/publishers
- Change the report name to 'NYPL DRB Total Item Requests by Title' (Sheet 1), 'NYPL DRB Total Item Requests by Title / Downloads' (Sheet 2), 'NYPL DRB Total Item Requests by Title / Views' (Sheet 3,  'NYPL DRB Total Item Requests by Title by Country' (Sheet 4
- Add an additional row with 'Metric_Type' (after Publisher_Name and before Reporting_Period) and add the metric description as suggested in the file
- Zeroed out rows for titles that were neither viewed nor downloaded in report period

The publishers' additional requests (i.e. moving the "Reporting Total" column, adding the full publisher name) were addressed in [this PR](https://github.com/NYPL/drb-etl-pipeline/pull/377).

### Issues while coding
This entire process was essentially trial-and-error -- Grabbing the DRB data in a time-efficient manner proved difficult given the `editions` and `records` tables share no foreign key, plus both tables have limited indexing. For the normalized data, I initially thought of hashing with a dictionary, but switched over to DataFrames to favor speed (which will prove useful for the large prod logs we'll be working with). I also ended up having to completely restructure the way we used `Pollers` (and ultimately did away with the `DownloadDataPoller` and `ViewDataPoller`). On top of all of this, figuring out how to implement zeroed out columns while not drastically impacting report generation performance was tricky -- I ended up doing some not-so-readable DF operations to optimize. In summary, this took a while! For good reason :)

### CODE REVIEW

**Comment by fatimarahman** [comment]
### Performance improvements
- Minimizing calls to DRB database by pulling all publisher project-related data from the DRB db beforehand, normalizing said data to Counter 5 formatting, and hashing for easier search. This is all done before we start processing access logs and generating reports. 
- Aggregating relevant logs before generating reports

### Changes requested by business analysts/publishers
- Change the report name to 'NYPL DRB Total Item Requests by Title' (Sheet 1), 'NYPL DRB Total Item Requests by Title / Downloads' (Sheet 2), 'NYPL DRB Total Item Requests by Title / Views' (Sheet 3,  'NYPL DRB Total Item Requests by Title by Country' (Sheet 4
- Add an additional row with 'Metric_Type' (after Publisher_Name and before Reporting_Period) and add the metric description as suggested in the file
- Zeroed out rows for titles that were neither viewed nor downloaded in report period

The publishers' additional requests (i.e. moving the "Reporting Total" column, adding the full publisher name) were addressed in [this PR](https://github.com/NYPL/drb-etl-pipeline/pull/377).

### Issues while coding
This entire process was essentially trial-and-error -- Grabbing the DRB data in a time-efficient manner proved difficult given the `editions` and `records` tables share no foreign key, plus both tables have limited indexing. For the normalized data, I initially thought of hashing with a dictionary, but switched over to DataFrames to favor speed (which will prove useful for the large prod logs we'll be working with). I also ended up having to completely restructure the way we used `Pollers` (and ultimately did away with the `DownloadDataPoller` and `ViewDataPoller`). On top of all of this, figuring out how to implement zeroed out columns while not drastically impacting report generation performance was tricky -- I ended up doing some not-so-readable DF operations to optimize. In summary, this took a while! For good reason :)

**Comment by fatimarahman** [comment]
Also! I left the "TODO" comments in so people can see exactly how the performance enhancements were implemented, as well as commented out the log aggregation just to make testing faster. I'll undo these changes once I get some feedback!

File: .python-version
Comment (kylevillegas93) [inline .python-version:1]: Hmm - I'm not positive that the rest of the drb-etl-pipeline is on 3.12.0. Recommend removing this file for now.

File: config/example.yaml
Comment (kylevillegas93) [inline config/example.yaml:1]: Hmm - let's leave this one up!

File: analytics/upress_reporting/counter_5_controller.py [hunk 1]
```diff
@@ -1,74 +1,198 @@
...
+        aggregate_logs.aggregate_logs_in_period(
+            date_range=self.reporting_period,
+            s3_bucket=self.download_bucket,
+            s3_path=os.environ.get("DOWNLOAD_LOG_PATH", None),
+            regex=DOWNLOAD_FILE_ID_REGEX,
+            referrer_url=self.referrer_url,
+        )
+
+    def pull_publisher_project_data(self) -> CTE:
+        # Purpose of method is to minimize db calls
```
Comment (kylevillegas93) [inline analytics/upress_reporting/counter_5_controller.py:130]: Recommend removing this comment :)

File: analytics/upress_reporting/counter_5_controller.py [hunk 2]
```diff
@@ -1,74 +1,198 @@
...
+                                          publisher_project_records.c.has_part,
+                                          publisher_project_records.c.identifiers,
+                                          publisher_project_records.c.publisher_project_source,
+                                          publisher_project_records.c.subjects,
+                                          literal(False).label("accessed"))
+            .join(publisher_project_editions,
+                  publisher_project_records.c.uuid == publisher_project_editions.c.uuid)
+            .cte("merged_publisher_project_data")
+        )
+        return merged_publisher_project_data
```
Comment (kylevillegas93) [inline analytics/upress_reporting/counter_5_controller.py:165]: nit: newline between the query and the return statement

File: analytics/upress_reporting/counter_5_controller.py [hunk 3]
```diff
@@ -1,74 +1,198 @@
...
+            port=os.environ.get("POSTGRES_PORT", None),
+            db=os.environ.get("POSTGRES_NAME", None),
+        )
+        self.db_manager.generateEngine()
+        self.db_manager.createSession()
 
     def create_reports(self):
-        print("Generating Counter 5 reports...")
+        print("Generating Counter 5 reports...", datetime.now())
+        # TODO: aggregate log files -> * one for views and one for downloads (this is agnostic of publisher)
```
Comment (kylevillegas93) [inline analytics/upress_reporting/counter_5_controller.py:55]: Are these TODOs done? If so, let's remove the comments - see below as well.

File: analytics/upress_reporting/helpers/format_data.py
Comment (kylevillegas93) [inline analytics/upress_reporting/helpers/format_data.py:1]: nice! really like this helper class :)

File: analytics/upress_reporting/models/pollers/poller.py [hunk 4]
```diff
@@ -1,137 +1,118 @@
...
 
 IP_REGEX = r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}"
+REQUEST_REGEX = r"REST.GET.OBJECT "
+TIMESTAMP_REGEX = r"\[.+\]"
 
 
-class Poller(ABC):
-    def __init__(self, publisher, date_range):
-        self.publisher = publisher
+class Poller():
```
Comment (kylevillegas93) [inline analytics/upress_reporting/models/pollers/poller.py:13]: poller is generic. maybe UsageEventDataPoller is more explicit?

File: analytics/upress_reporting/counter_5_controller.py (hunk 3 above)
Comment (fatimarahman) [reply analytics/upress_reporting/counter_5_controller.py]: Yep, they're done! Just left them in to illustrate how I addressed our original plan for improving performance. They are removed now

File: analytics/upress_reporting/models/pollers/poller.py (hunk 4 above)
Comment (fatimarahman) [reply analytics/upress_reporting/models/pollers/poller.py]: I renamed to InteractionEventPoller to stay consistent with the data type name! UsageEventDataPoller seems a little confusing to me but lmk your thoughts

**Comment by fatimarahman** [comment]
[Link](https://docs.google.com/spreadsheets/d/1I3s2eBOFjabF2WyEEPCatQzZmzheNa0TVUBR3BE5W2I/edit?usp=sharing) to sample report!

I added the description in [this comment](https://github.com/NYPL/drb-etl-pipeline/pull/366#issuecomment-2406029747) but can move it up.

@kylevillegas93 Now it takes 5 min to generate a report for a month (a 2 min improvement) and 15 to generate a report for a quarter (a 3 min improvement). If you leave out aggregating the logs, it takes a minute and a half to process and generate the actual report. I would say that for the future, we should look at speeding up aggregating the logs as well as dealing with max requests to http://ipinfo.io. This is the error I got today after generating a report:
`Status code 429 from http://ipinfo.io/108.27.223.104/json: ERROR - 429 Client Error: Too Many Requests for url: http://ipinfo.io/{ip_address}/json`

File: analytics/upress_reporting/models/pollers/poller.py (hunk 4 above)
Comment (kylevillegas93) [reply analytics/upress_reporting/models/pollers/poller.py]: that works!

**Comment by kylevillegas93** [comment]
looks good to me - approving here since I can't approve a PR I created ✅

**Pull request merged**